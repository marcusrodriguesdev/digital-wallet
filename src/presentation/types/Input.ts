export type scaleTypes = 'small' | 'medium' | 'large'
