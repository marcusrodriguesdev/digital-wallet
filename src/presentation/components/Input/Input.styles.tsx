import styled from 'styled-components'

import { scaleTypes } from '../types/Input'

export const Container = styled.div`
  position: relative;
`

interface LabelProps {
  isFocused: boolean
  error: boolean
  scale: scaleTypes
}

export const Label = styled.a<LabelProps>`
  ${({ isFocused, theme, scale }) => {
    if (isFocused && scale === 'small') {
      return theme?.fonts?.overline ?? {}
    } else if (!isFocused && scale === 'small') {
      return theme?.fonts?.smallInfo ?? {}
    } else if (isFocused && scale === 'medium') {
      return theme?.fonts?.overline ?? {}
    } else if (!isFocused && scale === 'medium') {
      return theme?.fonts?.info ?? {}
    } else if (isFocused && scale === 'large') {
      return theme?.fonts?.smallInfo ?? {}
    } else if (!isFocused && scale === 'large') {
      return theme?.fonts?.bigInfo ?? {}
    }
  }}
  background: ${({ isFocused, theme }) => (isFocused ? theme?.colors?.background : 'transparent')};
  border-radius: 2xl;
  color: ${({ theme, error, isFocused }) =>
    error
      ? theme?.colors?.error ?? '#ff6666'
      : isFocused
      ? theme?.colors?.highlight ?? '#7f5af0'
      : theme?.colors?.main ?? '#fffffe'};
  cursor: text;
  left: ${({ isFocused, scale }) => {
    if (isFocused && scale === 'small') {
      return '4px'
    } else if (isFocused && scale === 'medium') {
      return '6px'
    } else if (isFocused && scale === 'large') {
      return '8px'
    } else if ((!isFocused && scale === 'small') || scale === 'medium' || scale === 'large') {
      return '12px'
    }
  }};
  padding: 0.5px 5px;
  position: absolute;
  top: ${({ isFocused, scale }) => (isFocused ? -9 : scale === 'small' ? 11 : scale === 'medium' ? 12 : 13)}px;
  transition: 0.2s ease all;
  z-index: 1000;
`

interface StyledInputProps extends LabelProps {
  secureText?: boolean
}

export const StyledInput = styled.input<StyledInputProps>`
  ${({ theme, scale: sized }) =>
    sized === 'small'
      ? theme?.fonts?.smallInfo ?? {}
      : sized === 'medium'
      ? theme?.fonts?.infoNarrow ?? {}
      : theme?.fonts?.bigInfo ?? {}};
  background-color: ${({ theme }) => theme?.colors?.background ?? '#16161a'};
  border: ${({ theme, error, isFocused }) =>
      error ? theme?.colors?.error ?? '#ff6666' : theme?.colors?.highlight ?? '#7f5af0'}
    1px solid;
  border-radius: 2px;
  color: ${({ theme }) => theme?.colors?.main ?? '#fffffe'};
  height: ${({ scale: sized }) => (sized === 'small' ? 40 : sized === 'medium' ? 45 : 50)}px;
  padding-left: 15px;
  padding-bottom: 2px;
  width: ${({ scale: sized }) => (sized === 'small' ? 240 : sized === 'medium' ? 270 : 300)}px;

  &:focus {
    outline: none;
  }
`

interface ButtonProps {
  scale: scaleTypes
}

export const Button = styled.button<ButtonProps>`
  background: transparent;
  border: none;
  cursor: pointer;
  position: relative;
  right: 35px;
  top: 9px;
  z-index: 1000;

  img {
    background: transparent;
  }

  &:hover {
    opacity: 0.8;
  }
`

export const HelpMessageContainer = styled.div``

interface HelpMessageProps {
  error?: boolean
}

export const HelpMessage = styled.p<HelpMessageProps>`
  ${({ theme }) => theme?.fonts?.caption ?? {}};
  color: ${({ theme, error }) => (error ? theme?.colors?.error ?? '#ff6666' : theme?.colors?.main ?? '#fffffe')};
  position: absolute;
`
