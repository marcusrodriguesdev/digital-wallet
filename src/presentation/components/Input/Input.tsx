import React, { useRef, useState } from 'react'
import { Assets } from '../../../presentation/theme/Assets'
import { scaleTypes } from '../types/Input'
import { Button, Container, HelpMessage, HelpMessageContainer, Label, StyledInput } from './Input.styles'

export interface InputProps {
  label: string
  value: string
  name: string
  type: string
  error?: boolean
  scale: scaleTypes
  secureText?: boolean
  onChange: (event: React.ChangeEvent<HTMLInputElement>) => void
  onChangeType?: () => void
}

/**
 * Input component
 *
 * @param label           The label of the input
 * @param value           The value of the input
 * @param name            The name of the input
 * @param type            The type of the input
 * @param error           The error of the input
 * @param scale           The scale of the input
 * @param secureText      The boolean to show or hide the password
 * @param onChange        The function to be called when the input changes
 * @param onChangeType    The function to be called when the input type changes
 */

const Input: React.FC<InputProps> = (props: InputProps) => {
  const { label, value, name, type, secureText, error, scale, onChange, onChangeType } = props

  const [isFocused, setIsFocused] = useState(!!value)
  const [textFocused, setTextFocused] = useState(false)
  const inputRef = useRef({} as HTMLInputElement)

  const toggleFocus = (): void => {
    if (!value) {
      setIsFocused(!isFocused)
    }
    setTextFocused(!textFocused)
  }

  return (
    <Container>
      {label && (
        <Label
          scale={scale}
          isFocused={isFocused}
          error={error}
          data-testid="Input/Label"
          onClick={() => inputRef.current.focus()}>
          {label}
        </Label>
      )}
      <StyledInput
        secureText={secureText}
        isFocused={isFocused}
        name={name}
        onFocus={() => toggleFocus()}
        onBlur={() => toggleFocus()}
        ref={inputRef}
        value={value}
        type={type}
        error={error}
        onChange={onChange}
        data-testid="Input/StyledInput"
        scale={scale}
      />

      {secureText && type === 'text' && (
        <Button scale={scale} data-testid="Input/ButtonEyeOpen" type="button" onClick={onChangeType}>
          <img src={Assets.EyeOpen} alt="Eye open" />
        </Button>
      )}
      {secureText && type === 'password' && (
        <Button scale={scale} data-testid="Input/ButtonEyeClose" type="button" onClick={onChangeType}>
          <img className="eye-close" src={Assets.EyeClose} alt="Eye close" />
        </Button>
      )}
      {error && (
        <HelpMessageContainer data-testid="Input/Error">
          {secureText && <HelpMessage error={error}>Email ou senha incorreto</HelpMessage>}
        </HelpMessageContainer>
      )}
    </Container>
  )
}

export { Input }
