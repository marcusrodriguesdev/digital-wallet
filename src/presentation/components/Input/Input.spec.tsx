import React from 'react'
import { Input, InputProps } from './Input'
import { fireEvent, render } from '@testing-library/react'
import 'jest-styled-components'
import * as Faker from 'faker'

const makeSUT = (props?: Partial<InputProps>): any => {
  const value = props?.value ?? Faker.random.words(2)
  const name = props?.value ?? Faker.random.word()
  const label = props?.label ?? Faker.random.words(2)
  const type = props?.type ?? 'text'
  const error = props?.error ?? Faker.random.boolean()
  const secureText = props?.secureText ?? Faker.random.boolean()
  const scale = props?.scale ?? 'medium'
  const onChangeSpy = jest.fn()
  const onChangeTypeSpy = jest.fn()

  const sut = (
    <Input
      value={value}
      label={label}
      name={name}
      type={type}
      secureText={secureText}
      scale={scale}
      error={error}
      onChange={onChangeSpy}
      onChangeType={onChangeTypeSpy}
    />
  )

  return { sut, value, label, secureText, name, type, scale, error, onChangeSpy, onChangeTypeSpy }
}

describe('presentation/components/Input', () => {
  it('should match snapshot', () => {
    const { sut } = makeSUT({
      value: 'email@email.com',
      name: 'email',
      label: 'Email',
      type: 'text',
      secureText: false,
      error: false,
      scale: 'medium'
    })
    const inputRender = render(sut)
    expect(inputRender).toMatchSnapshot()
  })

  it('should call onChangeSpy with the typed text as argument when text change', () => {
    const { sut, onChangeSpy } = makeSUT()

    const { getByTestId } = render(sut)
    const input = getByTestId('Input/StyledInput')

    const changedText = Faker.random.word
    fireEvent.change(input, { target: { value: changedText } })

    expect(onChangeSpy).toBeCalledTimes(1)
  })
  it('should contain correct value', () => {
    const { sut, value } = makeSUT()
    const { getByTestId } = render(sut)
    const input = getByTestId('Input/StyledInput') as HTMLInputElement

    expect(input.value).toEqual(value)
  })

  it('should call onChangeTypeSpy when click on eye open button', () => {
    const { sut, onChangeTypeSpy } = makeSUT({
      secureText: true,
      type: 'text'
    })

    const { getByTestId } = render(sut)
    const button = getByTestId('Input/ButtonEyeOpen') as HTMLButtonElement

    fireEvent.click(button)

    expect(onChangeTypeSpy).toBeCalledTimes(1)
  })

  it('should call onChangeTypeSpy when click on eye close button', () => {
    const { sut, onChangeTypeSpy } = makeSUT({
      type: 'password',
      secureText: true
    })

    const { getByTestId } = render(sut)
    const button = getByTestId('Input/ButtonEyeClose') as HTMLButtonElement

    fireEvent.click(button)

    expect(onChangeTypeSpy).toBeCalledTimes(1)
  })

  it('should not contain eye open button when secureText is false', () => {
    const { sut } = makeSUT({
      secureText: false
    })

    const { queryByTestId } = render(sut)
    const button = queryByTestId('Input/ButtonEyeOpen')

    expect(button).toBeNull()
  })

  it('should contain correct label', () => {
    const { sut, label } = makeSUT()

    const { getByTestId } = render(sut)
    const labelElement = getByTestId('Input/Label')

    expect(labelElement.textContent).toEqual(label)
  })

  it('should contain correct name', () => {
    const { sut, name } = makeSUT()

    const { getByTestId } = render(sut)
    const input = getByTestId('Input/StyledInput') as HTMLInputElement

    expect(input.name).toEqual(name)
  })

  it('should contain correct type', () => {
    const { sut, type } = makeSUT()

    const { getByTestId } = render(sut)
    const input = getByTestId('Input/StyledInput') as HTMLInputElement

    expect(input.type).toEqual(type)
  })

  it('should contain error when error is true', () => {
    const { sut } = makeSUT({
      error: true
    })

    const { getByTestId } = render(sut)
    const error = getByTestId('Input/Error')

    expect(error).toBeTruthy()
  })

  it('should not contain error when error is false', () => {
    const { sut } = makeSUT({
      error: false
    })

    const { queryByTestId } = render(sut)
    const error = queryByTestId('Input/Error') as HTMLInputElement

    expect(error).toBeNull()
  })

  it('should has focus when label is clicked', () => {
    const { sut } = makeSUT()

    const { getByTestId } = render(sut)
    const label = getByTestId('Input/Label')

    fireEvent.click(label)
    const input = getByTestId('Input/StyledInput') as HTMLInputElement

    expect(document.activeElement).toBe(input)
  })

  it('should be in blur when value is empty', () => {
    const { sut } = makeSUT({
      value: ''
    })

    const { getByTestId } = render(sut)
    const label = getByTestId('Input/StyledInput')

    fireEvent.blur(label)

    expect(document.activeElement).toBe(document.body)
  })
})
