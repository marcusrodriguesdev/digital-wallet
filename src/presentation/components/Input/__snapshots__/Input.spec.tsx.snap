// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`presentation/components/Input should match snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": .c0 {
  margin-top: 20px;
  position: relative;
  margin-left: 20px;
}

.c1 {
  border-radius: 2xl;
  color: #7f5af0;
  cursor: text;
  left: 6px;
  padding: 0.5px 5px;
  position: absolute;
  top: -9px;
  -webkit-transition: 0.2s ease all;
  transition: 0.2s ease all;
  z-index: 1000;
}

.c2 {
  background-color: #16161a;
  border: #7f5af0 1px solid;
  border-radius: 2px;
  color: #fffffe;
  height: 45px;
  padding-left: 15px;
  padding-bottom: 2px;
  width: 270px;
}

.c2:focus {
  outline: none;
}

<body>
    <div>
      <div
        class="c0"
      >
        <a
          class="c1"
          data-testid="Input/Label"
          scale="medium"
        >
          Email
        </a>
        <input
          class="c2"
          data-testid="Input/StyledInput"
          name="email@email.com"
          scale="medium"
          type="text"
          value="email@email.com"
        />
      </div>
    </div>
  </body>,
  "container": .c0 {
  margin-top: 20px;
  position: relative;
  margin-left: 20px;
}

.c1 {
  border-radius: 2xl;
  color: #7f5af0;
  cursor: text;
  left: 6px;
  padding: 0.5px 5px;
  position: absolute;
  top: -9px;
  -webkit-transition: 0.2s ease all;
  transition: 0.2s ease all;
  z-index: 1000;
}

.c2 {
  background-color: #16161a;
  border: #7f5af0 1px solid;
  border-radius: 2px;
  color: #fffffe;
  height: 45px;
  padding-left: 15px;
  padding-bottom: 2px;
  width: 270px;
}

.c2:focus {
  outline: none;
}

<div>
    <div
      class="c0"
    >
      <a
        class="c1"
        data-testid="Input/Label"
        scale="medium"
      >
        Email
      </a>
      <input
        class="c2"
        data-testid="Input/StyledInput"
        name="email@email.com"
        scale="medium"
        type="text"
        value="email@email.com"
      />
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
